File: src/tacky_print.fs
module Tacky_print

open Tacky

// -- Helpers to replace OCaml's Format module --

let inline write (out: System.IO.TextWriter) (s: string) = out.Write(s)

let pp_print_list (pp_sep: System.IO.TextWriter -> unit)
                  (pp_item: System.IO.TextWriter -> 'a -> unit)
                  (out: System.IO.TextWriter)
                  (items: 'a list) =
    items |> List.iteri (fun i item ->
        if i > 0 then pp_sep out
        pp_item out item)

let comma_sep (out: System.IO.TextWriter) = write out ", "

let pp_init_list (out: System.IO.TextWriter) (init_list: Initializers.static_init list) =
    write out "{"
    pp_print_list comma_sep Initializers.pp_static_init out init_list
    write out "}"

let pp_unary_operator (out: System.IO.TextWriter) = function
    | Complement -> write out "~"
    | Negate -> write out "-"
    | Not -> write out "!"

// optional escape_brackets argument lets us escape < and > when instructions
// appear in HTML-style tables in graphviz
let pp_binary_operator (escape_brackets: bool) (out: System.IO.TextWriter) = function
    | Add -> write out "+"
    | Subtract -> write out "-"
    | Multiply -> write out "*"
    | Divide -> write out "/"
    | Mod -> write out "%"
    | Equal -> write out "=="
    | NotEqual -> write out "!="
    | LessThan ->
        let s = if escape_brackets then "&lt;" else "<"
        write out s
    | LessOrEqual ->
        let s = if escape_brackets then "&lt;=" else "<="
        write out s
    | GreaterThan ->
        let s = if escape_brackets then "&gt;" else ">"
        write out s
    | GreaterOrEqual ->
        let s = if escape_brackets then "&gt;=" else ">="
        write out s

let const_to_string = function
    | Const.ConstInt i -> sprintf "%d" i
    | Const.ConstLong l -> sprintf "%d" l + "l"
    | Const.ConstUInt ui -> sprintf "%u" ui + "u"
    | Const.ConstULong ul -> sprintf "%u" ul + "ul"
    | Const.ConstDouble d -> sprintf "%g" d
    | Const.ConstChar c -> sprintf "%d" c
    | Const.ConstUChar uc -> sprintf "%u" uc

let pp_tacky_val (out: System.IO.TextWriter) = function
    | Constant i -> write out (const_to_string i)
    | Var s -> write out s

let pp_tacky_val_list (out: System.IO.TextWriter) (vals: TackyVal list) =
    pp_print_list comma_sep pp_tacky_val out vals

let pp_string_list (out: System.IO.TextWriter) (strs: string list) =
    pp_print_list comma_sep (fun o s -> write o s) out strs

let pp_instruction (escape_brackets: bool) (out: System.IO.TextWriter) = function
    | Return None -> write out "Return"
    | Return (Some v) ->
        write out "Return("
        pp_tacky_val out v
        write out ")"
    | Unary { op = op; src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = "
        pp_unary_operator out op
        pp_tacky_val out src
    | Binary { op = op; src1 = src1; src2 = src2; dst = dst } ->
        pp_tacky_val out dst
        write out " = "
        pp_tacky_val out src1
        write out " "
        pp_binary_operator escape_brackets out op
        write out " "
        pp_tacky_val out src2
    | Copy { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = "
        pp_tacky_val out src
    | Jump s ->
        write out (sprintf "Jump(%s)" s)
    | JumpIfZero (cond, target) ->
        write out "JumpIfZero("
        pp_tacky_val out cond
        write out (sprintf ", %s)" target)
    | JumpIfNotZero (cond, target) ->
        write out "JumpIfNotZero("
        pp_tacky_val out cond
        write out (sprintf ", %s)" target)
    | Label s ->
        out.WriteLine()
        write out (sprintf "%s:" s)
    | FunCall { f = f; args = args; dst = None } ->
        write out (sprintf "%s(" f)
        pp_tacky_val_list out args
        write out ")"
    | FunCall { f = f; args = args; dst = Some dst } ->
        pp_tacky_val out dst
        write out (sprintf " = %s(" f)
        pp_tacky_val_list out args
        write out ")"
    | SignExtend { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = SignExtend("
        pp_tacky_val out src
        write out ")"
    | ZeroExtend { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = ZeroExtend("
        pp_tacky_val out src
        write out ")"
    | Truncate { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = Truncate("
        pp_tacky_val out src
        write out ")"
    | DoubleToInt { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = DoubleToInt("
        pp_tacky_val out src
        write out ")"
    | DoubleToUInt { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = DoubleToUInt("
        pp_tacky_val out src
        write out ")"
    | IntToDouble { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = IntToDouble("
        pp_tacky_val out src
        write out ")"
    | UIntToDouble { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = UIntToDouble("
        pp_tacky_val out src
        write out ")"
    | GetAddress { src = src; dst = dst } ->
        pp_tacky_val out dst
        write out " = GetAddress("
        pp_tacky_val out src
        write out ")"
    | Load v ->
        pp_tacky_val out v.dst
        write out " = Load("
        pp_tacky_val out v.src_ptr
        write out ")"
    | Store v ->
        write out "*("
        pp_tacky_val out v.dst_ptr
        write out ") = "
        pp_tacky_val out v.src
    | AddPtr { ptr = ptr; index = index; scale = scale; dst = dst } ->
        pp_tacky_val out dst
        write out " = "
        pp_tacky_val out ptr
        write out " + "
        pp_tacky_val out index
        write out (sprintf " * %d" scale)
    | CopyToOffset { src = src; dst = dst; offset = offset } ->
        write out (sprintf "%s[offset = %d] = " dst offset)
        pp_tacky_val out src
    | CopyFromOffset { src = src; offset = offset; dst = dst } ->
        pp_tacky_val out dst
        write out (sprintf " = %s[offset = %d]" src offset)

let pp_function_definition (escape_brackets: bool) (``global``: bool) (name: string)
                           (``params``: string list) (out: System.IO.TextWriter)
                           (body: TackyInstruction list) =
    if ``global`` then write out "global "
    write out (sprintf "%s(" name)
    pp_string_list out ``params``
    write out "):"
    out.WriteLine()
    write out "    "
    pp_print_list
        (fun o -> o.WriteLine(); write o "    ")
        (pp_instruction escape_brackets)
        out body

let pp_tl (escape_brackets: bool) (out: System.IO.TextWriter) = function
    | Function { name = name; ``global`` = ``global``; ``params`` = ``params``; body = body } ->
        pp_function_definition escape_brackets ``global`` name ``params`` out body
    | StaticVariable { ``global`` = ``global``; name = name; init = init; t = t } ->
        if ``global`` then write out "global "
        Types.pp out t
        write out (sprintf " %s = " name)
        pp_init_list out init
    | StaticConstant { name = name; init = init; t = t } ->
        write out "const "
        Types.pp out t
        write out (sprintf " %s = " name)
        Initializers.pp_static_init out init

let pp_program (escape_brackets: bool) (out: System.IO.TextWriter) (Program tls) =
    pp_print_list
        (fun o -> o.WriteLine(); o.WriteLine())
        (pp_tl escape_brackets)
        out tls
    out.WriteLine()
    out.Flush()

let debug_print_tacky (src_filename: string) tacky_prog =
    if Settings.Debug.Value then
        let tacky_file =
            UniqueIds.makeLabel (System.IO.Path.GetFileNameWithoutExtension(src_filename))
            + ".debug.tacky"
        let chan = new System.IO.StreamWriter(tacky_file)
        pp_program false (chan :> System.IO.TextWriter) tacky_prog
        chan.Close()

File: src/TackyPrint.rs
//! Pretty-printing for the TACKY intermediate representation.
#![allow(non_snake_case)]

use std::fs::File;
use std::io::{self, BufWriter, Write};
use std::path::Path;

use super::Const::ConstValue;
use super::Initializers::StaticInit;
use super::Tacky::*;
use super::Types::Type;
use super::UniqueIds::{self, Counter};

pub fn ppUnaryOperator(out: &mut impl Write, op: TackyUnaryOperator) -> io::Result<()> {
    write!(out, "{}", match op { TackyUnaryOperator::Complement => "~", TackyUnaryOperator::Negate => "-", TackyUnaryOperator::Not => "!" })
}

pub fn ppBinaryOperator(escape_brackets: bool, out: &mut impl Write, op: TackyBinaryOperator) -> io::Result<()> {
    let text = match op {
        TackyBinaryOperator::Add => "+", TackyBinaryOperator::Subtract => "-",
        TackyBinaryOperator::Multiply => "*", TackyBinaryOperator::Divide => "/",
        TackyBinaryOperator::Mod => "%", TackyBinaryOperator::Equal => "==",
        TackyBinaryOperator::NotEqual => "!=",
        TackyBinaryOperator::LessThan if escape_brackets => "&lt;",
        TackyBinaryOperator::LessOrEqual if escape_brackets => "&lt;=",
        TackyBinaryOperator::GreaterThan if escape_brackets => "&gt;",
        TackyBinaryOperator::GreaterOrEqual if escape_brackets => "&gt;=",
        TackyBinaryOperator::LessThan => "<", TackyBinaryOperator::LessOrEqual => "<=",
        TackyBinaryOperator::GreaterThan => ">", TackyBinaryOperator::GreaterOrEqual => ">=",
    };
    out.write_all(text.as_bytes())
}

pub fn constToString(value: ConstValue) -> String {
    match value {
        ConstValue::Int(v) => v.to_string(), ConstValue::Long(v) => format!("{v}l"),
        ConstValue::UInt(v) => format!("{v}u"), ConstValue::ULong(v) => format!("{v}ul"),
        ConstValue::Double(v) => format!("{v}"), ConstValue::Char(v) => v.to_string(),
        ConstValue::UChar(v) => v.to_string(),
    }
}

pub fn ppTackyVal(out: &mut (impl Write + ?Sized), value: &TackyVal) -> io::Result<()> {
    match value { TackyVal::Constant(c) => write!(out, "{}", constToString(*c)), TackyVal::Var(v) => out.write_all(v.as_bytes()) }
}

fn commaList<T>(out: &mut impl Write, values: &[T], mut pp: impl FnMut(&mut dyn Write, &T) -> io::Result<()>) -> io::Result<()> {
    for (index, value) in values.iter().enumerate() { if index != 0 { write!(out, ", ")?; } pp(out, value)?; }
    Ok(())
}

fn showType(t: &Type) -> String {
    match t {
        Type::Char => "Char".into(), Type::SChar => "SChar".into(), Type::UChar => "UChar".into(),
        Type::Int => "Int".into(), Type::Long => "Long".into(), Type::UInt => "UInt".into(),
        Type::ULong => "ULong".into(), Type::Double => "Double".into(), Type::Void => "Void".into(),
        Type::Pointer(inner) => format!("{}*", showType(inner)),
        Type::Array(inner, size) => format!("({}, {})", showType(inner), size),
        Type::Function(params, ret) => format!("(FunType (param_types = [{}], ret_type = {}))", params.iter().map(showType).collect::<Vec<_>>().join("; "), showType(ret)),
        Type::Struct(tag) => format!("(Structure {tag})"),
    }
}

fn showInit(init: &StaticInit) -> String {
    match init {
        StaticInit::CharInit(v) => v.to_string(), StaticInit::UCharInit(v) => v.to_string(),
        StaticInit::IntInit(v) => v.to_string(), StaticInit::LongInit(v) => format!("{v}l"),
        StaticInit::UIntInit(v) => format!("{v}u"), StaticInit::ULongInit(v) => format!("{v}ul"),
        StaticInit::DoubleInit(v) => v.to_string(), StaticInit::ZeroInit(v) => format!("zero[{v}]"),
        StaticInit::StringInit(s, nul) => format!("\"{s}{}\"", if *nul { "\\0" } else { "" }),
        StaticInit::PointerInit(s) => format!("&{s}"),
    }
}

pub fn ppInstruction(escape_brackets: bool, out: &mut impl Write, instruction: &TackyInstruction) -> io::Result<()> {
    use TackyInstruction::*;
    match instruction {
        Return(None) => write!(out, "Return"), Return(Some(v)) => { write!(out, "Return(")?; ppTackyVal(out, v)?; write!(out, ")") }
        Unary(v) => { ppTackyVal(out, &v.dst)?; write!(out, " = ")?; ppUnaryOperator(out, v.op)?; ppTackyVal(out, &v.src) }
        Binary(v) => { ppTackyVal(out, &v.dst)?; write!(out, " = ")?; ppTackyVal(out, &v.src1)?; write!(out, " ")?; ppBinaryOperator(escape_brackets, out, v.op)?; write!(out, " ")?; ppTackyVal(out, &v.src2) }
        Copy(v) => assignment(out, &v.dst, "", &v.src),
        Jump(s) => write!(out, "Jump({s})"), JumpIfZero(v, s) => conditional(out, "JumpIfZero", v, s),
        JumpIfNotZero(v, s) => conditional(out, "JumpIfNotZero", v, s), Label(s) => write!(out, "\n{s}:"),
        FunCall(v) => { if let Some(dst) = &v.dst { ppTackyVal(out, dst)?; write!(out, " = ")?; } write!(out, "{}(", v.f)?; commaList(out, &v.args, |o, x| ppTackyVal(o, x))?; write!(out, ")") }
        SignExtend(v) => conversion(out, "SignExtend", v), ZeroExtend(v) => conversion(out, "ZeroExtend", v),
        Truncate(v) => conversion(out, "Truncate", v), DoubleToInt(v) => conversion(out, "DoubleToInt", v),
        DoubleToUInt(v) => conversion(out, "DoubleToUInt", v), IntToDouble(v) => conversion(out, "IntToDouble", v),
        UIntToDouble(v) => conversion(out, "UIntToDouble", v), GetAddress(v) => conversion(out, "GetAddress", v),
        Load(v) => { ppTackyVal(out, &v.dst)?; write!(out, " = Load(")?; ppTackyVal(out, &v.src_ptr)?; write!(out, ")") }
        Store(v) => { write!(out, "*(")?; ppTackyVal(out, &v.dst_ptr)?; write!(out, ") = ")?; ppTackyVal(out, &v.src) }
        AddPtr(v) => { ppTackyVal(out, &v.dst)?; write!(out, " = ")?; ppTackyVal(out, &v.ptr)?; write!(out, " + ")?; ppTackyVal(out, &v.index)?; write!(out, " * {}", v.scale) }
        CopyToOffset(v) => { write!(out, "{}[offset = {}] = ", v.dst, v.offset)?; ppTackyVal(out, &v.src) }
        CopyFromOffset(v) => { ppTackyVal(out, &v.dst)?; write!(out, " = {}[offset = {}]", v.src, v.offset) }
    }
}

fn assignment(out: &mut impl Write, dst: &TackyVal, prefix: &str, src: &TackyVal) -> io::Result<()> { ppTackyVal(out, dst)?; write!(out, " = {prefix}")?; ppTackyVal(out, src) }
fn conditional(out: &mut impl Write, name: &str, value: &TackyVal, target: &str) -> io::Result<()> { write!(out, "{name}(")?; ppTackyVal(out, value)?; write!(out, ", {target})") }
fn conversion(out: &mut impl Write, name: &str, value: &TackySrcDst) -> io::Result<()> { ppTackyVal(out, &value.dst)?; write!(out, " = {name}(")?; ppTackyVal(out, &value.src)?; write!(out, ")") }

pub fn ppFunctionDefinition(escape_brackets: bool, global: bool, name: &str, params: &[String], out: &mut impl Write, body: &[TackyInstruction]) -> io::Result<()> {
    if global { write!(out, "global ")?; } write!(out, "{name}(")?;
    commaList(out, params, |o, s| o.write_all(s.as_bytes()))?; writeln!(out, "):")?; write!(out, "    ")?;
    for (i, instruction) in body.iter().enumerate() { if i != 0 { write!(out, "\n    ")?; } ppInstruction(escape_brackets, out, instruction)?; }
    Ok(())
}

pub fn ppTl(escape_brackets: bool, out: &mut impl Write, item: &TackyTopLevel) -> io::Result<()> {
    match item {
        TackyTopLevel::Function(f) => ppFunctionDefinition(escape_brackets, f.global, &f.name, &f.params, out, &f.body),
        TackyTopLevel::StaticVariable(v) => { if v.global { write!(out, "global ")?; } write!(out, "{} {} = {{", showType(&v.t), v.name)?; commaList(out, &v.init, |o, x| write!(o, "{}", showInit(x)))?; write!(out, "}}") }
        TackyTopLevel::StaticConstant(v) => write!(out, "const {} {} = {}", showType(&v.t), v.name, showInit(&v.init)),
    }
}

pub fn ppProgram(escape_brackets: bool, out: &mut impl Write, program: &TackyProgram) -> io::Result<()> {
    for (i, item) in program.0.iter().enumerate() { if i != 0 { write!(out, "\n\n")?; } ppTl(escape_brackets, out, item)?; }
    writeln!(out)?; out.flush()
}

pub fn debugPrintTacky(debug: bool, counter: Counter, src_filename: &Path, program: &TackyProgram) -> io::Result<Counter> {
    if !debug { return Ok(counter); }
    let stem = src_filename.file_stem().and_then(|s| s.to_str()).unwrap_or_default();
    let (counter, label) = UniqueIds::make_label(stem, counter);
    let mut file = BufWriter::new(File::create(format!("{label}.debug.tacky"))?);
    ppProgram(false, &mut file, program)?;
    Ok(counter)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn prints_program_in_original_debug_format() {
        let program = TackyProgram(vec![TackyTopLevel::Function(TackyFunctionDef { name: "main".into(), global: true, params: vec![], body: vec![TackyInstruction::Binary(TackyBinaryInfo { op: TackyBinaryOperator::LessThan, src1: TackyVal::Constant(ConstValue::Int(1)), src2: TackyVal::Constant(ConstValue::Int(2)), dst: TackyVal::Var("tmp.0".into()) }), TackyInstruction::Return(Some(TackyVal::Var("tmp.0".into())))] })]);
        let mut out = Vec::new(); ppProgram(true, &mut out, &program).unwrap();
        assert_eq!(String::from_utf8(out).unwrap(), "global main():\n    tmp.0 = 1 &lt; 2\n    Return(tmp.0)\n");
    }
}
